File: case-study/Boiler Control/sim/src/telemetry_logger.cpp
#include "telemetry_logger.h"
#include <cmath>
#include <iomanip>
#include <iostream>
#include <stdexcept>

namespace boiler {

TelemetryLogger::TelemetryLogger(const std::string& filepath)
{
    file_.open(filepath);
    if (!file_.is_open())
        throw std::runtime_error("Cannot open log file: " + filepath);

    file_ << "t,y1,y2,y3,u1,u2,u3,du1,du2,du3,"
          << "ref_y1,ref_y2,ref_y3,e1,e2,e3,"
          << "IAE_y1,IAE_y2,IAE_y3,ISE_y1,ISE_y2,ISE_y3,E_valve\n";
    file_ << std::scientific << std::setprecision(6);
}

TelemetryLogger::~TelemetryLogger()
{
    if (file_.is_open()) file_.flush();
}

void TelemetryLogger::log(const TickData& d)
{
    if (first_) {
        t_prev_ = d.t;
        first_  = false;
    }
    double dt = (d.t > t_prev_ + 1e-9) ? (d.t - t_prev_) : Ts_;
    t_prev_ = d.t;

    double e1 = d.ref_y(0) - d.y(0);
    double e2 = d.ref_y(1) - d.y(1);
    double e3 = d.ref_y(2) - d.y(2);

    IAE_[0] += std::abs(e1) * dt;
    IAE_[1] += std::abs(e2) * dt;
    IAE_[2] += std::abs(e3) * dt;
    ISE_[0] += e1 * e1 * dt;
    ISE_[1] += e2 * e2 * dt;
    ISE_[2] += e3 * e3 * dt;
    E_valve_ += (d.du(0)*d.du(0) + d.du(1)*d.du(1) + d.du(2)*d.du(2)) * dt;

    file_ << d.t
          << ',' << d.y(0)     << ',' << d.y(1)     << ',' << d.y(2)
          << ',' << d.u(0)     << ',' << d.u(1)     << ',' << d.u(2)
          << ',' << d.du(0)    << ',' << d.du(1)    << ',' << d.du(2)
          << ',' << d.ref_y(0) << ',' << d.ref_y(1) << ',' << d.ref_y(2)
          << ',' << e1         << ',' << e2          << ',' << e3
          << ',' << IAE_[0]    << ',' << IAE_[1]    << ',' << IAE_[2]
          << ',' << ISE_[0]    << ',' << ISE_[1]    << ',' << ISE_[2]
          << ',' << E_valve_   << '\n';
}

void TelemetryLogger::flush()
{
    if (file_.is_open()) file_.flush();
    std::cout << std::fixed << std::setprecision(4)
              << "  IAE=[" << IAE_[0] << ", " << IAE_[1] << ", " << IAE_[2] << "]"
              << "  ISE=[" << ISE_[0] << ", " << ISE_[1] << ", " << ISE_[2] << "]"
              << "  E_valve=" << E_valve_ << "\n";
}

} // namespace boiler

File: case-study/Boiler Control MATLAB/matlab/+boiler/TelemetryLogger.m
classdef TelemetryLogger < handle
% TELEMETRYLOGGER  Per-tick CSV writer + IAE/ISE/E_valve accumulator.
%   Emits the exact 23-column schema of the C++ telemetry_logger so the output
%   flows through the existing Python report pipeline unchanged:
%     t,y1,y2,y3,u1,u2,u3,du1,du2,du3,ref_y1,ref_y2,ref_y3,e1,e2,e3,
%     IAE_y1,IAE_y2,IAE_y3,ISE_y1,ISE_y2,ISE_y3,E_valve
%   Error convention matches the C++ logger: e = ref - y, dt = Ts each step.

    properties (Access = private)
        fid = -1
        IAE = [0 0 0]
        ISE = [0 0 0]
        Ev  = 0
        Ts  = 1.0
    end

    methods
        function obj = TelemetryLogger(filepath)
            obj.fid = fopen(filepath, 'w');
            if obj.fid < 0
                error('boiler:log', 'Cannot open log file: %s', filepath);
            end
            fprintf(obj.fid, ['t,y1,y2,y3,u1,u2,u3,du1,du2,du3,', ...
                'ref_y1,ref_y2,ref_y3,e1,e2,e3,', ...
                'IAE_y1,IAE_y2,IAE_y3,ISE_y1,ISE_y2,ISE_y3,E_valve\n']);
        end

        function log(obj, t, y, u, du, ref)
            dt = obj.Ts;
            e  = ref(:) - y(:);
            obj.IAE = obj.IAE + abs(e).' * dt;
            obj.ISE = obj.ISE + (e.^2).' * dt;
            obj.Ev  = obj.Ev  + sum(du(:).^2) * dt;
            fprintf(obj.fid, ['%.6e,%.6e,%.6e,%.6e,%.6e,%.6e,%.6e,', ...
                '%.6e,%.6e,%.6e,%.6e,%.6e,%.6e,%.6e,%.6e,%.6e,', ...
                '%.6e,%.6e,%.6e,%.6e,%.6e,%.6e,%.6e\n'], ...
                t, y(1), y(2), y(3), u(1), u(2), u(3), ...
                du(1), du(2), du(3), ref(1), ref(2), ref(3), ...
                e(1), e(2), e(3), ...
                obj.IAE(1), obj.IAE(2), obj.IAE(3), ...
                obj.ISE(1), obj.ISE(2), obj.ISE(3), obj.Ev);
        end

        function [iae, ise, ev] = finals(obj)
            iae = obj.IAE;  ise = obj.ISE;  ev = obj.Ev;
        end

        function close(obj)
            if obj.fid >= 0, fclose(obj.fid); obj.fid = -1; end
        end

        function delete(obj)
            obj.close();
        end
    end
end
